packages/core-contracts/src/contracts/arks/MoonwellArk.sol

```diff
@@ -0,0 +1,251 @@
+            .getAllMarketConfigs(address(mToken));
+
+        rewardTokens = new address[](marketConfigs.length);
+        rewardAmounts = new uint256[](marketConfigs.length);
+        for (uint256 i = 0; i < marketConfigs.length; i++) {
+            rewardTokens[i] = marketConfigs[i].emissionToken;
+            uint256 rewardBalance = IERC20(marketConfigs[i].emissionToken)
+                .balanceOf(address(this));
+            rewardAmounts[i] = rewardBalance;
+            IERC20(marketConfigs[i].emissionToken).safeTransfer(
+                _raft,
+                rewardBalance
+            );
+        }
+    }
+
+    /**
+     * @notice Validates the board data
+     * @dev This Ark does not require any validation for board data
+     * @param /// data Additional data to validate (unused in this implementation)
+     */
+    function _validateBoardData(bytes calldata) internal override {}
+
+    /**
+     * @notice Validates the disembark data
+     * @dev This Ark does not require any validation for disembark data
+     * @param /// data Additional data to validate (unused in this implementation)
+     */
+    function _validateDisembarkData(bytes calldata) internal override {}
+
+    /**
+     * @notice Get the underlying balance of a user with accrued interest, without modifying state
+     * @param user The address of the user to check
+     * @return The amount of underlying tokens the user effectively owns, including accrued interest
+     */
+    function balanceOfUnderlyingWithInterest(
+        address user
+    ) public view returns (uint256) {
+        (, uint256 shares, , uint256 storedExchangeRate) = mToken
+            .getAccountSnapshot(user);
+
+        uint256 accrualBlockTimestampPrior = mToken.accrualBlockTimestamp();
+
+        // If no time has passed, use stored rate
+        if (accrualBlockTimestampPrior == block.timestamp) {
+            return (shares * storedExchangeRate) / 1e18;
+        }
+
+        uint256 exchangeRate = _calculateCurrentExchangeRate();
+        return (shares * exchangeRate) / 1e18;
+    }
+
+    function _calculateCurrentExchangeRate() internal view returns (uint256) {
+        uint256 totalSupply = mToken.totalSupply();
+        if (totalSupply == 0) {
+            return mToken.exchangeRateStored();
+        }
+
+        uint256 totalCash = mToken.getCash();
+        uint256 borrowsPrior = mToken.totalBorrows();
+        uint256 reservesPrior = mToken.totalReserves();
+
+        uint256 interestAccumulated = _calculateInterestAccumulated(
+            totalCash,
+            borrowsPrior,
+            reservesPrior
+        );
+
+        uint256 totalReserves = _calculateNewReserves(
+            reservesPrior,
+            interestAccumulated
+        );
+        uint256 totalBorrows = interestAccumulated + borrowsPrior;
+
+        uint256 _totalAssets = totalCash + totalBorrows - totalReserves;
+        return _totalAssets.divWadDown(totalSupply);
```

Comment:
I believe you need scale by 1e18 before dividing, as getCash will return values in the underlying decimals, which could be e6 for USDC, for example. Please run some tests to validate this as I'm not 100% sure